File: 排序算法/merge_sort.py
# 归并排序-自顶向下递归

def merge(left, right):
    result = []
    while left and right:  # 两部分的元素作比较
        if left[0] <= right[0]:
            result.append(left.pop(0))  # 小的先加入列表中
        else:
            result.append(right.pop(0))

    # 确保两部分的元素都已归并
    if left:
        result.extend(left)
    if right:
        result.extend(right)
    
    return result

def mergeSort(arr):
    if(len(arr)<2):
        return arr
    
    middle = len(arr) // 2
    left, right = arr[0: middle], arr[middle:]
    arr_left = mergeSort(left)  # 先对左半部分排序
    arr_right = mergeSort(right)  # 再对右半部分排序
    return merge(arr_left, arr_right)

arr = [22,33,49,47,33,12,68,29,22]
print(mergeSort(arr))

File: 排序算法/merge_sort.java
package com.sort;

/**
 * @ClassName mergeSort
 * @Description TODO
 * @Author Lijuce_K
 * @Date 2021/9/15 21:01
 * @Version 1.0
 **/
public class mergeSort {
    public static void main(String[] args) {
        int[] nums = {22,33,49,47,33,12,68,29,22};
        callSort(nums);
        for(int n: nums)
            System.out.println(n);
    }

    public static void callSort(int[] array) {
        mergeSort(array, 0, array.length-1);
    }

    public static void mergeSort(int[] array, int left, int right) {
        int mid = (left + right) / 2;
        if (left < right) {
            mergeSort(array, left, mid);
            mergeSort(array, mid+1, right);
            merge(array, left, mid, right);
        }
    }

    public static void merge(int[] array, int left, int mid, int right) {
        int[] temp = new int[right-left+1];
        int leftStartIdx = left;
        int rightStartIdx = mid+1;
        int idx = 0;
        while (leftStartIdx <= mid && rightStartIdx <= right) {
            if (array[leftStartIdx] < array[rightStartIdx]) {
                temp[idx++] = array[leftStartIdx++];
            } else {
                temp[idx++] = array[rightStartIdx++];
            }
        }
        while (leftStartIdx <= mid) {
            temp[idx++] = array[leftStartIdx++];
        }
        while (rightStartIdx <= right) {
            temp[idx++] = array[rightStartIdx++];
        }
        for (int i = 0; i < idx; i++) {
            array[left+i] = temp[i];
        }
    }
}
